### DISCUSSION TITLE
Yearly partitioned asset where the latest year gets updated every day

### DISCUSSION THREAD

**Discussion by EtienneT** [dagster-io/dagster#14612] [Q&A] [answered]
I have a daily partionned asset that gets updated every day called price_daily.

Then I would like a new asset based on price_daily that would generate one file per year, let's say price_yearly.  Ideally I would use a yearly partionned asset but there doesn't seem to have a partition def for that.  Those files are used in some downstream processes to easily load multiple years in ML tasks.

Every day I want the year in progress to get re-generated, but all past years should be left alone.  The current year would depend on all price_daily of the current year.

Should I use @assets, or I need something else?

**Comment by sryza** [comment, marked as answer]
You can define a yearly partitions definition with:

```python
from dagster import TimeWindowPartitionsDefinition

yearly_partitions_def = TimeWindowPartitionsDefinition(
    cron_schedule="0 0 1 1 *",
    fmt="%Y-%m-%d",
    start="2020-01-01",
    end_offset=1,
)
```

Note the `end_offset=1`. Without this, the `PartitionsDefinition` will only include years that have already concluded. If I understand correctly, you want to also include the current, un-concluded year.

Then you can make a yearly-partitioned asset that depends on a daily-partitioned asset:

```python
from dagster import DailyPartitionsDefinition, asset


@asset(partitions_def=DailyPartitionsDefinition(start_date="2020-01-01"))
def daily_asset() -> None:
    ...


@asset(
    partitions_def=yearly_partitions_def,
    non_argument_deps={"daily_asset"},
)
def yearly_asset() -> None:
    ...
```

Finally, write a sensor that triggers the yearly asset whenever there's a materialization of the daily-partitioned asset:

```python
from datetime import datetime
from dagster import AssetSelection, Definitions, RunRequest, multi_asset_sensor


@multi_asset_sensor(
    monitored_assets=AssetSelection.assets(daily_asset),
    request_assets=AssetSelection.assets(yearly_asset),
)
def my_sensor(context):
    record = context.latest_materialization_records_by_key()[daily_asset.key]
    if record is not None:
        daily_partition = record.partition_key
        date = datetime.strptime(daily_partition, daily_asset.partitions_def.fmt)
        context.advance_all_cursors()
        return RunRequest(partition_key=f"{date.year}-01-01")


defs = Definitions(assets=[daily_asset, yearly_asset], sensors=[my_sensor])
```

**Reply by EtienneT** [reply]
Wow thanks for the great answer!!

**Comment by sryza** [comment]
Related: https://github.com/dagster-io/dagster/discussions/18893
